plugins/cxaudio/fu-cxaudio-device.c

```diff
@@ -0,0 +1,881 @@
+					  error)) {
+		g_prefix_error (error, "failed to read EEPROM custom info: ");
+		return FALSE;
+	}
+	if (cinfo.LayoutSignature == FU_CXAUDIO_SIGNATURE_BYTE)
+		self->eeprom_layout_version = cinfo.LayoutVersion;
+	g_debug ("CpxPatchVersion: %u.%u.%u",
+		 cinfo.CpxPatchVersion[0],
+		 cinfo.CpxPatchVersion[1],
+		 cinfo.CpxPatchVersion[2]);
+	g_debug ("SpxPatchVersion: %u.%u.%u.%u",
+		 cinfo.SpxPatchVersion[0],
+		 cinfo.SpxPatchVersion[1],
+		 cinfo.SpxPatchVersion[2],
+		 cinfo.SpxPatchVersion[3]);
+	g_debug ("VendorID: 0x%04x", cinfo.VendorID);
+	g_debug ("ProductID: 0x%04x", cinfo.ProductID);
+	g_debug ("RevisionID: 0x%04x", cinfo.RevisionID);
+	g_debug ("ApplicationStatus: 0x%02x", cinfo.ApplicationStatus);
+
+	/* serial number, which also allows us to recover it after write */
+	if (self->eeprom_layout_version >= 0x01) {
+		self->serial_number_set = cinfo.SerialNumberStringAddress != 0x0;
+		if (self->serial_number_set) {
+			g_autofree gchar *tmp = NULL;
+			tmp = fu_cxaudio_device_eeprom_read_string (self,
+								    cinfo.SerialNumberStringAddress,
+								    error);
+			if (tmp == NULL)
+				return FALSE;
+			fu_device_set_serial (device, tmp);
+		}
+	}
+
+	/* read fw version */
+	if (!fu_cxaudio_device_operation (self,
+					  FU_CXAUDIO_OPERATION_READ,
+					  FU_CXAUDIO_MEM_KIND_CPX_RAM,
+					  FU_CXAUDIO_REG_FIRMWARE_VERSION_ADDR,
+					  verbuf_fw, sizeof(verbuf_fw),
+					  FU_CXAUDIO_OPERATION_FLAG_NONE,
+					  error)) {
+		g_prefix_error (error, "failed to read EEPROM firmware version: ");
+		return FALSE;
+	}
+	version_fw = g_strdup_printf ("%02X.%02X.%02X.%02X",
+				      verbuf_fw[1], verbuf_fw[0],
+				      verbuf_fw[3], verbuf_fw[2]);
+	fu_device_set_version_bootloader (device, version_fw);
+
+	/* use a different address if a patch is in use */
+	if (self->eeprom_patch_valid_addr != 0x0) {
+		if (!fu_cxaudio_device_ensure_patch_level (self, error))
+			return FALSE;
+	}
+	addr = self->patch_level == 0 ? FU_CXAUDIO_EEPROM_CPX_PATCH_VERSION_ADDRESS :
+					FU_CXAUDIO_EEPROM_CPX_PATCH2_VERSION_ADDRESS;
+	if (!fu_cxaudio_device_operation (self,
+					  FU_CXAUDIO_OPERATION_READ,
+					  FU_CXAUDIO_MEM_KIND_EEPROM,
+					  addr, verbuf_patch, sizeof(verbuf_patch),
+					  FU_CXAUDIO_OPERATION_FLAG_NONE,
+					  error)) {
+		g_prefix_error (error, "failed to read EEPROM patch version: ");
+		return FALSE;
+	}
+	version_patch = g_strdup_printf ("%02X-%02X-%02X",
+					 verbuf_patch[0], verbuf_patch[1], verbuf_patch[2]);
+	fu_device_set_version (device, version_patch, FWUPD_VERSION_FORMAT_PLAIN);
+
+	/* find out if patch supports additional capabilities (optional) */
+	cap_str = g_usb_device_get_string_descriptor (usb_device,
+						      FU_CXAUDIO_DEVICE_CAPABILITIES_STRIDX,
+						      NULL);
+	if (cap_str != NULL) {
+		g_auto(GStrv) split = g_strsplit (cap_str, ";", -1);
+		for (guint i = 0; split[i] != NULL; i++) {
+			if (g_strcmp0 (split[i], "RESET") == 0)
+				self->sw_reset_supported = TRUE;
+		}
+	}
```

Comment:
Any other useful capabilities worth putting in debug log while reading this?

Reply:
I think it's just RESET, but have added a debug too for me future self.